### PR TITLE
Classify discards as identifier

### PR DESCRIPTION
**Pull request by jcouv**
Adds classification for discard designation syntax, and identifier syntax represents a discard symbol.

@CyrusNajmabadi @roslyn-ide 
Two questions:
- Is `Identifier` the correct classification for discards?
- Although I verified the classification in unittest and also by debugging the dev hive, the colors still seem wrong. Any tips on how to troubleshoot that?

image

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
yeah, those should def not be blue...

File: src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/DiscardDesignationSyntaxClassifier.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+{
+    internal class DiscardDesignationSyntaxClassifier : AbstractSyntaxClassifier
+    {
+        public override void AddClassifications(
+            SyntaxNode syntax,
+            SemanticModel semanticModel,
+            ArrayBuilder<ClassifiedSpan> result,
+            CancellationToken cancellationToken)
+        {
+            if (syntax is DiscardDesignationSyntax discardDesignation)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/DiscardDesignationSyntaxClassifier.cs]: isn't this redundant?  i.e. shoudln't this just be a straight cast given that you only register for DiscardDesignationSyntax?

**Comment by CyrusNajmabadi** [comment]
> Although I verified the classification in unittest and also by debugging the dev hive, the colors still seem wrong. Any tips on how to troubleshoot that?

That's super weird... you may need to set breakpoints in teh syntactic and semantic classification taggers and validate that *all* the spans they're returning are correect, and no identifier classificaiton is getting overridden by some other classification.

**Comment by DustinCampbell** [comment]
Out of curiosity, why do we need a separate classifier to handle purely-syntactic classification? I would expect purely-syntactic classification to be done in https://github.com/dotnet/roslyn/blob/master/src/Workspaces/CSharp/Portable/Classification/ClassificationHelpers.cs.

`ISyntaxClassifier` has a misleading name because it's generally used for semantic classification.

**Comment by jcouv** [comment]
@DustinCampbell I didn't know about ClassificationHelpers.cs ;-) Let me try that approach.

File: src/Workspaces/CSharp/Portable/Classification/ClassificationHelpers.cs [hunk 2]
```diff
@@ -24,6 +23,10 @@ internal static class ClassificationHelpers
         /// <returns>The correct syntactic classification for the token.</returns>
         public static string GetClassification(SyntaxToken token)
         {
+            if (token.IsKind(SyntaxKind.DiscardDesignation, SyntaxKind.UnderscoreToken))
```
Comment (DustinCampbell) [inline src/Workspaces/CSharp/Portable/Classification/ClassificationHelpers.cs]: So much smaller! :smile:

**Comment by jcouv** [comment]
Moving the logic to ClassificationHelpers seems to have solved the problem in IDE. The problem was likely some problem of classifications getting overridden.

Now the IDE looks ok, although I still wonder whether Identifier is the proper classification. The problem is that it doesn't distinguish between underscore (name of a variable) and discard.

image

**Comment by DustinCampbell** [comment]
> Now the IDE looks ok, although I still wonder whether Identifier is the proper classification. The problem is that it doesn't distinguish between underscore (name of a variable) and discard.

I suppose, but that's probably OK. What else would you expect the classification to be? (Note that creating new classifications is a lot of work.)

**Comment by DustinCampbell** [comment]
FWIW, I think it's OK that the user can't distinguish by color the difference between a variable named underscore and a discard. In general, I expect most users that have variables named underscore are already using it in a way that is spiritually the same as a discard. If I'm correct, having them be different would add little value.

**Comment by Pilchie** [comment]
I agree with @DustinCampbell here.

**Comment by jcouv** [comment]
Sounds good to me. Thanks

**Comment by jcouv** [comment]
test windows_debug_vs-integration_prtest  please

**Comment by jcouv** [comment]
It's unclear what is the VS integration error. ([details](https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_vs-integration_prtest/5766/))
I'll re-run.

```
18:34:07 Running 1 test assemblies in 1 partitions
18:34:14    1 running,  0 queued,  0 completed
18:59:18    0 running,  0 queued,  1 completed,  1 failures
18:59:18 Errors Roslyn.VisualStudio.IntegrationTests.dll: 
18:59:18 
18:59:18 Command: C:\Users\dotnet-bot\.nuget\packages\xunit.runner.console\2.2.0-beta4-build3444\tools\xunit.console.x86.exe "D:\j\workspace\windows_debug---0c191de0\Binaries\Debug\UnitTests\VisualStudioIntegrationTests\Roslyn.VisualStudio.IntegrationTests.dll"  -xml "D:\j\workspace\windows_debug---0c191de0\Binaries\Debug\UnitTests\VisualStudioIntegrationTests\xUnitResults\Roslyn.VisualStudio.IntegrationTests.dll.xml" -noshadow -verbose
18:59:18 xUnit output log: D:\j\workspace\windows_debug---0c191de0\Binaries\Debug\UnitTests\VisualStudioIntegrationTests\xUnitResults\Roslyn.VisualStudio.IntegrationTests.dll.out.log
18:59:18 
xunit produced no error output but had exit code 1
18:59:18 ================
```

**Comment by jcouv** [comment]
test windows_debug_vs-integration_prtest please

**Comment by jcouv** [comment]
test windows_debug_vs-integration_prtest please

**Comment by CyrusNajmabadi** [comment]
👍

**Pull request merged**